### DISCUSSION TITLE
Ban by default attaching sockets to all network interfaces

### DISCUSSION THREAD

**Discussion by balloob** [home-assistant/architecture#509] [General]
# Context

Some integrations use broadcast messages to receive updates from devices on the network. If the host system has multiple network interfaces, attaching a socket to listen for these broadcasts might be bound to the wrong interface. This causes messages to get lost. This is an edge case and most users are not impacted.

To work around this, some integrations will iterate over all available interfaces and try to bind to all of them. This can cause issues with some network interfaces. The most important one that causes issues is `local loopback` but users could also have virtual interfaces for Docker etc. 

We've found multiple ways of trying to filter out the interfaces that we shouldn't listen to. But none of these are the magic bullet that work perfect in every use case.

Some related issues:
 - https://github.com/home-assistant/core/issues/45957
 - https://github.com/home-assistant-libs/aioshelly/issues/68 

# Proposal

New rule: All code that listens for broadcasts in Home Assistant should listen to the default network interface or the specified network interfaces. No code should listen to all network interfaces. 

We can then add a new feature to Home Assistant to allow the user to specify the interfaces that should be listened to if the default does not suffice.

# Consequences

From a quick scan, the following integrations will be impacted and should be updated accordingly:

 - Shelly
 - Sonos
 - Zeroconf
 - SSDP

**Comment by dgomes** [comment]
"This is an edge case and most users are not impacted." yet this might be a breaking change for setup where HA lies between 2 different networks (e.g. Home network and IoT network)

I think this should be something handled by the integration or keep the current behaviour unless explicitly configured.

**Reply by frenck** [reply]
Shouldn't those cases resolve that on a network level? E.g., by setting up an mdns repeater/reflector? IGMP-proxies?

**Reply by chemelli74** [reply]
Sorry but I don't think it can be solved at network level. 
If you have 2 interfaces and you bind only to the default one, there is no way to get messages from the second. The only option is to bind to both.

**Reply by frenck** [reply]
I think you are now referring to a different use case than set in the openings post?
As for the broadcast traffic types set in the start topic (zeroconf/ssdp/sonos and such) that is not true.

**Reply by chemelli74** [reply]
Humm, not sure. 
When you need to receive multicast packets, you need to do a "IP_ADD_MEMBERSHIP" and this should iterate on all interfaces you want to receive traffic from.
This was the origin of the broken Shelly integration in the last main release. Fix was to avoid bindings in the time been.

**Reply by frenck** [reply]
Aah, but the problem is putting HA on multiple networks at the same time. Gotcha.
But that is what I meant in the first place, resolve that on a network level, so Home Assistant doesn't need to be on multiple networks. And even if you have multiple networks, route that traffic correctly across (from<->to) where needed selectively.

**Reply by thecode** [reply]
> Aah, but the problem is putting HA on multiple networks at the same time. Gotcha.
> But that is what I meant in the first place, resolve that on a network level, so Home Assistant doesn't need to be on multiple networks. And even if you have multiple networks, route that traffic correctly across (from<->to) where needed selectively.

I agree with you that the right solution is solve this using correct routing.
My personal opinion is that for now at least HA should not support more than one network interface. However since HA is considered to be a solution which keeps security and privacy at the top priority it would be nice to have a longer term discussion of how to support multiple interfaces. In addition since supervised Debian installations are also official (TBH I moved from such installation to hassos when first it was declared non official) these installations might require users tweaking to select the right interface (an example can be a wired network for devices and a cellular modem for external access.

A longer term solution might address different configurations such as:
1. Single interface (today)
2. Multiple interfaces, one for local/IOT access and one for cloud only access.
3. Mixed configurations.

Anyhow I absolutely support the current decision, I would like to suggest to add a checkbox when submitting a request and ticking the "new integration" checkbox, to add another checkbox under it "If my integration is using network access it uses the default network interface" (or something similar). I tried to look what is the correct status today, but many integration uses external packages and it is impossible to review all of them. Attaching to network interfaces incorrectly may not only cause stability issues, but also security issues if the integration design is faulty.

**Reply by dgomes** [reply]
IP Multicast is handled per interface and not per address, that is what's wrong with the linked issues. When you happen to have multiple addresses in the same interface you get those errors. 

Code subscribing to Multicast traffic should be interface wise and not address wise.

You should list interfaces and iterate that list, not list all addresses.

**Reply by thecode** [reply]
This is only part of the issue, there are also interfaces which should not be added. Currently integrations which join multicast groups on all interfaces may cause problems. Until there is a solution to specify which interfaces integrations should used, I think integrations should not be allowed to bind to specific interface and only use the default interface as suggested in the topic.

**Reply by dgomes** [reply]
"**may** cause problems" - can you provide an example of such problems?

The linked issues are an error caused by misuse of the IP Multicast API, the proper action is to not subscribe multiple times through the same interface.

**Reply by amelchio** [reply]
Those two linked issues are the same? I agree with @dgomes, this looks like a single bug more than a generic issue.

The Sonos integration has been scanning all network interfaces for two years and I don't recall reports of any issue with that. Scanning the wrong network, however, was previously enough of an issue that the interface enumeration was added.

Core support for specifying bind addresses sounds nice but it would probably need per-integration overrides. There are a lot of ~~crazy~~ advanced network setups out there.

**Reply by chemelli74** [reply]
I don't think we should handle all possible network scenarios, but at least the most common.

I have a single interface on HA, and do all the routing/firewalling outside, but having 2 interfaces ( one for internet and one for local communication) is not that unusual and I think would be a great addition for the project.

Just my 2 cents,

Simone

**Comment by balloob** [comment]
I've been doing some more research. And I think that we shouldn't even allow configuring multiple network interfaces to begin with. If we look at the network stack, it's not Home Assistant's problem to solve it. It should be solved either at the network level that is providing the multiple interfaces ([tutorial for Unifi](Best Practices for Managing Chromecast/Google Home on UniFi Network](https://help.ui.com/hc/en-us/articles/360001004034-UniFi-Best-Practices-for-Managing-Chromecast-Google-Home-on-UniFi-Network)) or at the host machine ([HassOS solution](https://github.com/home-assistant/plugin-multicast/blob/master/rootfs/etc/services.d/mdns/run)).

If people decide to run complicated setups, we should not try to patch over some of the problems we encounter. It's better for people to see it fail and know something is wrong and fix it.

**Reply by amelchio** [reply]
My setup with two network interfaces is to avoid opening the `http` integration to my IoT VLAN. Should that be considered a bad configuration?

**Comment by felipediel** [comment]
Hello everyone.

I don't like the idea of banning multiple network interfaces because Home Assistant is a hub. The main purpose is to connect things. The more networks it can access, the more things it can connect and the better hub it will be. It looks like a regression compared to what we have today.

Delegating the netif decision to the end user also doesn't seem like a good idea if we want to make Home Assistant more user friendly. Many people don't know what a network interface is and they are not interested in the learning curve. They want to "plug and play" and I think they are right. I can predict issues like "devices cannot be discovered" or "updates failed" in which we will have to troubleshoot/explain how to select an appropriate interface. We can antecipate ourselves and use this effort to build something that just works.

We went through a similar problem in the Broadlink integration and this is the solution we came to:

Do not bind() before broadcasting and do not use 255.255.255.255 as the broadcast address. Instead, send the message to the correct L3 broadcast address (e.g. 192.168.0.255) and leave for the OS to bind. This works like a charm in our synchoronous context, there is no sidecase or local loopback risk.

Of course, in some cases bind() will be necessary, especially when we are dealing with asynchronous communication. The solution in such cases is to select the local L3 IP addresses and bind correctly. This filter should do the trick:

```python3
import psutil

nics = psutil.net_if_addrs()

local_ip_addrs = [
    addr.address
    for addrs in nics.values()
    for addr in addrs
    if (
        addr.broadcast is not None
        and addr.family == socket.AF_INET
        and addr.netmask == "255.255.255.0"
    )
]

for addr in local_ip_addrs:
    # Your logic, e.g. s.bind((addr, port))
```

I think we can move part of this logic to the core, as it can be reused by many integrations, how about?

Edit: The above solution is for broadcasting. What matters here is the L3 filter which also works for multicasting.

**Reply by balloob** [reply]
But why would it be Home Assistant's responsibility? If it's the user who splits their network in two, they can also set it up to route the right traffic to the interface Home Assistant is on. We need to try to solve problems as close to where the problem is caused. In this case it's the network layer. Not the application layer.

**Reply by amelchio** [reply]
> We need to try to solve problems as close to where the problem is caused.

It is not a problem per se, it is just a complex situation – a situation that the application layer often cares about:

* https://nginx.org/en/docs/http/ngx_http_core_module.html#listen
* https://httpd.apache.org/docs/2.4/mod/mpm_common.html#listen
* http://www.postfix.org/postconf.5.html#inet_interfaces
* https://doc.dovecot.org/configuration_manual/service_configuration/#inet-listeners
* https://dev.mysql.com/doc/refman/8.0/en/server-system-variables.html#sysvar_bind_address
* https://github.com/memcached/memcached/wiki/ConfiguringServer#networking

... and here is the official Sonos client scanning both of my interfaces simultaneously during startup:

![Screen Shot 2021-02-12 at 01 58 11](https://user-images.githubusercontent.com/26174915/107718333-fccc1880-6cd5-11eb-9525-011e66443d49.png)

**Reply by Adminiuga** [reply]
> a situation that the application layer often cares 

Correct, but that does not mean the application decides on what interfaces to listen. In most cases, by default it listens on all interfaces, by binding to `INADDR_ANY` which is quite different from the application iterating over all interfaces and binding each one individually.
I'm proponent of  the "admin" being in control on what interfaces the application should listen/bind, not the application trying to outsmart itself.

**Reply by felipediel** [reply]
> But why would it be Home Assistant's responsibility? If it's the user who splits their network in two, they can also set it up to route the right traffic to the interface Home Assistant is on. We need to try to solve problems as close to where the problem is caused. In this case it's the network layer. Not the application layer.

As a user, I think connecting devices from different networks is a valid use case for a home automation hub. Apart from superior technology, what differentiates a Z-Wave stick from an extra Wi-Fi interface? Wouldn't be nice if we could take advantage of all the networking features of the machine and make the complexity transparent to the user?

From an architectural point of view, I think you are absolutely right not to want this as default, at least for now. This is a sidecase, it brings complexity, most integrations don't support this feature and the default must be something that works.

But could we at least make this optional and provide the local IP addresses via hass object to help integrations improve support gradually? Apart from the problems I mentioned above, there is nothing wrong with broadcasting or multicasting on multiple interfaces if done correctly.

**Reply by thecode** [reply]
@felipediel I think your comment summarize what most of us (or at least me) think: it makes sense to support multiple interfaces, this should not be the default and we should think of a way to provide integrations with a list of interfaces to use.

**Comment by agners** [comment]
Unfortunately the socket networking API is rather complex, I'd like to get a bit more clarity what we try to address here. I tried to put this together as best as I can but if I got it wrong or miss something, please help me out.

First off, there are generally different addressing styles:

- **Unicasting**: point-to-point, the typical network communication. The only communication 
- **Broadcasting**: point-to-all, requires raw IP or UDP frames (not supported in IPv6!)
- **Multicasting**: point-to-"a set" requires raw IP or UDP frames

(there is also anycasting, but doesn't really has relevance for HA)

In general, the socket API tries to abstract away interface selection by using IP addresses instead.

To listen to Unicast and Broadcast messages no network interface needs to be specified. By binding to a specific IP, the operating system uses the IP configuration to understand what interfaces to listen on (or rather, it only accepts packages on a interface with the correct address). E.g. binding to `192.168.1.10` makes sure to only listen on that interface. Listening for `192.168.1.255` listens for broadcasts. There is also the option to listen on  any IP address (and therefor interface) using `INADDR_ANY` (or `0.0.0.0`).

I think listening to any interface in the unicast and broadcast mode is not really a problem. In fact that is what a lot of servers are doing, including the HA http integration in its default configuration.

But, from what I understand, this issue is about **multicast**, e.g. as used by aioshelly, in turn used by the Shelly integration (https://github.com/home-assistant-libs/aioshelly/blob/0.6.0/aioshelly/coap.py#L41).

For Multicast, there is a socket option `IP_MULTICAST_IF` which allows to specify the interface for **outgoing** multicasts. If not specified, the behavior seems OS specific. "UNIX Network Programming" says:
> Berkley-derived kernels choose the default interface for an outgoing multicast datagram by searching the normal IP routing table for a route to the destination multicast address, and the corresponding interface is used. This is the same technique used to choose the receiving interface if the process does not specify one when joining a group. The assumption is that if a route exists for a given multicast address (perhaps the default route in the routing table), then the resulting interface should be used for input and output."

[The Linux Documentation Project](https://tldp.org/HOWTO/Multicast-HOWTO-6.html) says for `IP_MULTICAST_IF`
> Usually, the system administrator specifies the default interface multicast datagrams should be sent from. The programmer can override this and choose a concrete outgoing interface for a given socket with this option. 

From what I understand "selecting the interface" on Linux means adding something like this to the routing table e.g. `ip route add multicast 224.0.1.0/24 dev eth1`. So I think Linux basically does what is stated above for "Berkley-derived kernels".

**Reply by agners** [reply]
IMHO, not specifying the interface lets the OS choose the right interface basically defers the problem to whoever maintains the OS. With this method only one interface can be supported, but that seems a sensible default behavior. This also supports the use case of a dedicated IoT network e.g. using VLANs or a dedicated interface. The administrator just needs to setup the routing table correctly.

Simply for looping over all interfaces and multicast on all of them seems like a bad idea to me. Banning that for multicast seems sensible to me.

If really multiple interfaces are required, then doing it properly with a list of interfaces maintained centrally seems sensible.

**Reply by chemelli74** [reply]
> IMHO, not specifying the interface lets the OS choose the right interface basically defers the problem to whoever maintains the OS. 

Unfortunately to  be able to **receive** multicast packets, you need an explicit `socket.IP_ADD_MEMBERSHIP` thus forcing you to decide what to listen for.
If you leave out the interface, OS will listen only to first interface found (= lower index), making it working fine only for systems with one interface.

**Reply by Adminiuga** [reply]
in other words, users have to setup their network for multicast correctly and it is not HA's job to workaround the misconfiguration (or no configuration in 99%) by adding each interface into the multicast group and listening/binding on each interface.

**Reply by chemelli74** [reply]
Is at application layer that you decide if to join a multicast group or not.
There is no way for a final user to manually join the group.

**Reply by dgomes** [reply]
The default should always be to listen for **multicast** in **all** interfaces.

Look at avahi-daemon for instance, by default it listens to **all** interface except loopback and ptp, and exceptionally you can define a list of interfaces.

**Reply by Adminiuga** [reply]
yes. And the application should do it only on the  **default** interface. you don't have to add **all** your interfaces to the multicast group.

**Reply by chemelli74** [reply]
Why ? What is the reason behind your sentence ? 

If a system has 2 or more interfaces, is not misconfiguration, but a design/architecture choice.
And in such scenario, you have to bind to all interfaces that may carry traffic, not arbitrary says that it must be all routed by default interface. This is an assumption from the application side that is often wrong in real life scenarios.

So there are only 2 choices:

- state that only 1 network interface is supported in HA, and force every user to design their network based on this limitation.
- support multiple network interfaces, and fix application to handle such scenarios.

As a network engineer, I'm more than happy to understand real points against such.

Simone

**Reply by Adminiuga** [reply]
A system having two network interfaces is **not** a misconfiguration. A network that does not support "multicast" (i.e. forwarding multicast packets to the subscribed parties across broadcast domains/vlans) is a ~~misconfiguration~~ limitation, which does require a work around of:
1. system to be multihomed
2. App binding the socket on each interface

**Reply by elupus** [reply]
It's not a misconfiguration to have two interfaces. You can't change the enumeration order of interfaces in most os:s. So can't be solved on host level.

You need to be explicit when listening to multicast. That is just how it works. This applies both to ipv6 and ipv4. And or link local ipv6 where it becomes even more important.

For link local ipv6, you HAVE to specify scope id (interface) when listening and transmitting, otherwise there is no way for the os to know what you are doing.

Most w10 machines will for example broadcast their upnp server (if enabled) on link local ipv6.

**Comment by bdraco** [comment]
I opened a PR to switch `ssdp` to use `async_upnp_client` for the scans https://github.com/home-assistant/core/pull/46554. I'd appreciate testing/input from someone who has a bifurcated network with multiple interfaces.

**Comment by chemelli74** [comment]
For all integration but zeroconf, we may use the incoming ip of the device to know the right binding:

```python
ipr = IPRoute()
ip_for_ha_bind = ipr.route("get", dst=ip_from_zeroconf_packet)[0].get_attrs("RTA_PREFSRC")[0]
```

This will work only if all devices from same integration are on the same network or use the same routing interface.

Simone

**Reply by amelchio** [reply]
That seems to put the cart before the horse, there is no incoming IP address before we listen.

**Reply by chemelli74** [reply]
> That seems to put the cart before the horse, there is no incoming IP address before we listen.

Not really, if you read carefully I wrote that this is valid for all integration but zeroconf.
That way you have the incoming packet from zeroconf that can be used to get the info for binding, thus receiving real data ;-)

Simone

**Comment by amelchio** [comment]
I think this discussion is all over the place and should be closed. Anohter one can be opened where the issues and proposals are more clear.

Here are some points that have me wondering what we are actually discussing:
* sending vs. receiving broadcasts
* weasel wording like "can cause issues"
* assuming everything is HAOS
* assuming everything is Zeroconf
* assuming everything can be proxied

**Reply by felipediel** [reply]
I think this discussion can be more productive if:

1. We accept that each integration has its own way of communicating. We should avoid creating universal rules that will restrict their functioning or demand a refactor without clear benefits. If it's not broken, we don't fix it. If it's broken, we fix the problem where it is.

2. We accept that dealing with multiple network interfaces is something interesting and used by many people. It doesn't have to be mandatory, but it should also not be prohibited.

3. We discuss a way for Home Assistant to support the integrations with the name of the interfaces and the local IP addresses so that they don't miss the shot.

When I mentioned broadcasting it was not with the intention of distorting the topic, I was just exposing how we solved a similar problem by getting all available interfaces and applying a L3 filter, a solution that is suitable for all integrations. Not all will use the filter in the same way. Some will use the local IP addresses to bind(), some will send the message to the L3 broadcast addresses and leave for the OS to bind(), some will bind() to all interfaces and filter after, some will need the name of the interface for low level socket operations etc. But one thing they have in common is: they can use the same filter to do it correctly.

I think this is a decent filter, do we need anything else?
```python3
>>> import psutil
>>> nics = psutil.net_if_addrs()
>>> cards = [
    (netif, addr.address)
    for netif, addrs in nics.items()
    for addr in addrs
    if (
        addr.family == socket.AF_INET
        and addr.netmask == "255.255.255.0"
    )
]
>>> cards

[("wlxe894f6275302", "192.168.31.100"), ("enp3s0", "192.168.0.16")]
```

Obtaining the interfaces is a blocking operation, it would be better if we could do it only once. I think this is something to be done at the core and shared with the integrations via hass object or via observer pattern (in case we want to allow the user to enable/disable interfaces in the UI).

**Reply by amelchio** [reply]
Not all LANs have a /24 netmask.

But the sticking point is that while several solutions are being proposed, _nobody has described the problem_ (except that it "can cause issues").

**Reply by chemelli74** [reply]
There is a limit in the bind options for a socket. So when you exceed this limit, nothing works any more.

Systems with more interfaces, depending on the number of those, can fall in this situation easily.
To limit bind to useful interfaces there are many approaches, but none seems to be the real final fix for all scenarios.

Simone

**Reply by felipediel** [reply]
> Not all LANs have a /24 netmask.
> 
> But the sticking point is that while several solutions are being proposed, _nobody has described the problem_ (except that it "can cause issues").

The only problem I see here is: if all the integrations decide to support multiple network interfaces in their own way, we may end up with extra dependencies and duplicate code, and If at some point we decide to coordinate these efforts and offer more control over this in the UI (eg. enable/disable netifs), we will need a rework.

I think we can discuss a framework to help things grow clean. This is not urgent, not even necessary, but I think it is a valid point for discussion if we overcome the idea of banning multiple network interfaces.

And let's face it, building a filter that works well in all cases is not a simple task, but it is possible. We have a lot of geniuses here, I'm sure we'll come up with something that works. How do you think we can improve that filter? I don't discard user intervention, but I think we should at least eliminate `lo` and `docker0` for example, I see no reason to show these options.

**Reply by chemelli74** [reply]
I think I got it:

```python
from pyroute2 import IPRoute
import socket


def get_all_ips():
    """Get all ip from ethernet interfaces."""

    ip_list = []
    ip = IPRoute()

    for nic in ip.get_links(ip.link_lookup(operstate="UP")):
        iface = nic.get_attr("IFLA_IFNAME")
        iface_addr = ip.get_addr(family=socket.AF_INET, label=iface)
        ip_list.append(iface_addr[0].get_attr("IFA_ADDRESS"))
    ip.close()

    return ip_list


print(get_all_ips())
```

First filter for physical interfaces in state up ( so `lo` and `docker0` are filtered out), then look for IPv4 addresses ( and leave out IPv6).

Simone

**Reply by felipediel** [reply]
Works fine on my computer, but what if those interfaces are up?

**Reply by thecode** [reply]
The script above crashed on two hassio VMs (QEMU), some internal error from pyroute. This is why I usually don't like messing with low level network stuff. I know Simone tested it on many different Linux OS, but again the most trivial one did not work.
So the "Works fine for me" does not catch here :-)
I still think that this discussion is not focusing on the important topic. I understand everyone wants to suggest a solution or some does not like the "band by default" idea, but IMO the real topic here is if HA should support multiple interfaces and not how it should support it.
If the answer is yes, than this should be supported directly by HA core, same as `zeroconf`, `mqtt` and others are a common framework for integrations ,this should be here also. HA should build a list of interfaces, probably with some properties, such as if this interface is used for local access, cloud or anything which can help integrations to understand what to do with it.

**Reply by chemelli74** [reply]
> Works fine on my computer, but what if those interfaces are up?

In all my tests, only physical interfaces are "up" in pyroute

Simone

**Comment by chemelli74** [comment]
I made a small PR, [#47521](https://github.com/home-assistant/core/pull/47521), that will allow a wider usage of `get_local_ip()` that is already part of core.

Once you have the ip of the device, using this function you'll be able to know which is the ip of HA that will be used to start the connection.

Simone

**Reply by amelchio** [reply]
This discussion is about broadcasting on all interfaces, why would you do that if you have the IP of the device ...

**Reply by chemelli74** [reply]
> This discussion is about broadcasting on all interfaces, why would you do that if you have the IP of the device ...

This discussion is about attaching socket ( aka binding ).

Depends on device and protocols, but you may need binding only to receive updates, not to configure devices.

So if, and only if, devices have mDNS or can be configured manually, then with this code change you can bind to the right interface and only to that, without the need to bind to all.

I was not aware such function exists in core, found just yesterday.

Simone

**Reply by amelchio** [reply]
If you read the Context and the Proposal, it really is about broadcasts. But as I already said, I think this discussion is so confused that we should just give up and close it down without resolution.

**Comment by chemelli74** [comment]
Trying to come to a final decision, I'll ask again the main question: 

should HA support more than 1 network interface ?

IMHO, only after we have a definitive position on that, we can discuss about bindings.

Simone

**Comment by bdraco** [comment]
After changing `zeroconf` defaults to the using the default interface as the default, it solved a few loopback problems, but was a breaking change for other users. Using the default interface makes the most sense for the majority of user cases, and the analytics support this conclusion as most users are running operating system.

For a few versions `ssdp` has only been using the default interface when it was switched over to be async, and it hasn't generated any issues that I'm aware of.

When working on `sonos` recently, it is binding to all interfaces and broadcasting SSDP discovery on all of them.

`upnp` has a setting to specifically choose the ip to run discovery from. 

With each integration doing something a little bit different, the user has to figure out multiple different configuration for each integration that each present in a unique manner.

To solve this an avoid the original problem, I think it makes sense to provide a central configuration for this so each integration isn't reinventing this.

Proposed configuration options:

Auto Detect (default)
Bind to Default Interface
Bind to specific Interfaces (specify ips: [ X , X , X ] )
Bind to All Interfaces

Default: Use the auto detection logic currently used by `zeroconf` : https://github.com/home-assistant/core/blob/dev/homeassistant/components/zeroconf/__init__.py#L132

**Comment by felipediel** [comment]
This is how I see an ideal scenario:

1. Home Assistant provides integrations with a list of available interfaces (including name, family, netmask, broadcast addresses and local IP addresses). The first interface is the standard one, which will be used by integrations that do not support multiple interfaces.

2. Integrations can discover and communicate to devices on these networks. They can use `bind('', PORT)` for performance, but the message cannot be directed to unlisted networks. If they use multicast, they cannot add membership to IPs outside the range. If they use broadcast, they need to send the message to the available broadcast addresses. If they use unicast, they cannot send messages to IPs outside the range. There are helper functions to determine if an IP or broadcast address is allowed.

3. In the UI there are checkboxes in which the user can enable/disable interfaces. `lo`, `docker0` and anything that might cause issues is disabled by default. The rest is enabled. There is a drop-down menu in which the user can select the default network interface. The interface the server is on is the default.

**Comment by bdraco** [comment]
I went to start writing an actual implementation of this and realized almost nothing can consume it in the UI without a restart (no way to reload zeroconf, ssdp, etc) so this probably makes sense to be a yaml configuration as anything besides auto detect is going to be for advanced used anyways.

**Comment by jjlawren** [comment]
> I went to start writing an actual implementation of this and realized almost nothing can consume it in the UI without a restart (no way to reload zeroconf, ssdp, etc) so this probably makes sense to be a yaml configuration as anything besides auto detect is going to be for advanced used anyways.  

Even if a restart is required to take effect, a UI that shows available and selected interfaces _could_ still provide a better experience. If it's worth the effort when YAML suffices for advanced use is a different question.

**Reply by bdraco** [reply]
Something like: `Changes to this setting will only affect integrations that are loaded after it was changed, a restart may be required for integrations that are already loaded`.  Then we need to show what it was set to at boot

**Reply by bdraco** [reply]
Proposed UI here

https://github.com/home-assistant/frontend/pull/9210

Core `network` integration to support this

https://github.com/home-assistant/core/pull/50874

**Comment by frenck** [comment]
Closing this one stale for now. Most of this has been applied over the years anyway.

../Frenck
